### PR TITLE
Workspace updates

### PR DESCRIPTION
**Pull request by aciidb0mb3r**

### CODE REVIEW

**Comment by aciidb0mb3r** [comment]
@ddunbar this contains misc changes to Workspace, I can split these into multiple PRs if this becomes difficult to review

File: Sources/Commands/Workspace.swift [hunk 1]
```diff
@@ -468,6 +448,17 @@ public class Workspace {
...
     }
+
+    /// Load a manifest which is guaranteed to load because it has been loaded successfully previously.
+    /// A failure here indicates the manifest was changed somehow outside of SwiftPM.
+    // FIXME: Eventually get these manifests from a persistence cache.
+    func loadManifest(packagePath path: AbsolutePath, baseURL: String, version: Version?) throws -> Manifest {
+        do {
+            return try manifestLoader.load(packagePath: path, baseURL: baseURL, version: version)
+        } catch {
+            throw WorkspaceOperationError.unexpectedManifestLoadingError("Manifest at \(path.asString) unexpectedly failed to load. Try a clean build.")
```
Comment (aciidb0mb3r) [inline Sources/Commands/Workspace.swift]: I wrote clean build here considering clean will wipe out the checkouts (but keep the repositories)

File: Sources/Commands/Workspace.swift [hunk 2]
```diff
@@ -545,7 +532,9 @@ public class Workspace {
         data["version"] = .int(Workspace.currentSchemaVersion)
         data["dependencies"] = .array(dependencies.map{ $0.toJSON() })
 
-        // FIXME: This should write atomically.
-        try localFileSystem.writeFileContents(statePath, bytes: JSON.dictionary(data).toBytes())
+        // Write atomically.
+        let tempFile = try TemporaryFile()
+        try localFileSystem.writeFileContents(tempFile.path, bytes: JSON.dictionary(data).toBytes())
+        try rename(old: tempFile.path.asString, new: statePath.asString)
```
Comment (aciidb0mb3r) [inline Sources/Commands/Workspace.swift]: considering this is ok, should this be moved to Utility perhaps?

File: Sources/Basic/DictionaryExtensions.swift [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ This source file is part of the Swift.org open source project
+
+ Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+extension Dictionary {
```
Comment (ddunbar) [inline Sources/Basic/DictionaryExtensions.swift]: Let's put this in a separate PR, I think we all should review extensions to core types in Basic (partly so that we are all aware of which ones we add, and can use them consistently where appropriate).

Comment (ddunbar) [inline Sources/Basic/DictionaryExtensions.swift]: ... which reminds me: I suspect there are a number of places we could use this in places where we have the following pattern:
```
var foo = Dictionary<>()
... initialize foo ...
self.foo = foo
```

Using this init instead makes it obvious that everything is immutable post-initialization.

File: Sources/Commands/Workspace.swift [hunk 4]
```diff
@@ -494,35 +494,31 @@ public class Workspace {
         }
 
         // Load the state.
-        //
-        // FIXME: Build out improved file reading support.
-        try fopen(statePath) { handle in
-            let json = try JSON(bytes: ByteString(encodingAsUTF8: try handle.readFileContents()))
+        let json = try JSON(bytes: try localFileSystem.readFileContents(statePath))
```
Comment (ddunbar) [inline Sources/Commands/Workspace.swift]: Can we cleanup CheckoutManager to be consistent with this too? Really they should share common basic serialization support, but it didn't seem worth trying to design that right now.

File: Sources/Commands/Workspace.swift (hunk 2 above)
Comment (ddunbar) [inline Sources/Commands/Workspace.swift]: I think it is more likely it should move to part of the `FileSystem` API (as Foundation does). Correctly writing atomically is a very file system specific operation (the approach here actually isn't 100% correct on macOS in the case of immediate power loss, but the correct approach is much more expensive).

I would probably rather not fix this here (the risk of this bug being a major problem in the short term is low) and only try to fix it the "right" way.

File: Sources/Commands/Workspace.swift [hunk 5]
```diff
@@ -281,22 +284,15 @@ public class Workspace {
...
                 // If so, load its manifest.
-                //
-                // This should *never* fail, because we should only have ever
-                // got this checkout via loading its manifest successfully.
-                //
-                // FIXME: Nevertheless, we should handle this failure explicitly.
-                //
-                // FIXME: We should have a cache for this.
-                let manifest: Manifest = try! manifestLoader.load(packagePath: checkoutsPath.appending(managedDependency.subpath), baseURL: managedDependency.repository.url, version: managedDependency.currentVersion)
+                let manifest: Manifest = try loadManifest(packagePath: checkoutsPath.appending(managedDependency.subpath), baseURL: managedDependency.repository.url, version: managedDependency.currentVersion)
```
Comment (ddunbar) [inline Sources/Commands/Workspace.swift]: I think it would be better to fix this by trying to fix the core issue (which is that we should find a way to get the _actual_ manifest that was previously loaded).

The problem with cleaning this up is that IMHO it ends up obscuring that what the code is doing is actually *wrong*. I find that more obvious when the code has things like `try!` and `FIXMEs` littered around.

Note that the my comment about "we should have a cache for this" is probably wrong -- I think in this case what we want is to guarantee that we get the actual manifest that was read, not just happen to get the same manifest by going through a cache. This is a subtle, but important distinction: when the code doesn't need to use a cache to ensure it gets the same manifest then it ends up being more "solid" -- it becomes an invariant that this code operates on the same data that was previously seen, rather than an implicit dependency on the cache state.

File: Sources/Basic/DictionaryExtensions.swift (hunk 3 above)
Comment (aciidb0mb3r) [inline Sources/Basic/DictionaryExtensions.swift]: Opened #637

**Comment by ddunbar** [comment]
Sending back with some comments... I don't think we should take all of this in the current form. Let's break up the pieces and discuss individually.

File: Sources/Commands/Workspace.swift (hunk 4 above)
Comment (aciidb0mb3r) [inline Sources/Commands/Workspace.swift]: opened #638
